src/hotspot/share/classfile/stringTable.cpp

```diff
@@ -529,20 +531,26 @@ bool StringTable::do_rehash() {
   return true;
 }
 
+bool StringTable::should_grow() {
+  return get_load_factor() > PREF_AVG_LIST_LEN && !_local_table->is_max_size_reached();
+}
+
+uint StringTable::rehash_table_expected_workers() {
```

Comment:
We always reply either `0` or `1` from these. So logically, it seems to answer whether we need heavy work. So, this method is then simpler, and matching the structure of the relevant `rehash_table`:

```
// Quick checks if we would do any heavy-weight safepoint work in rehash_table.
// Add shortcut cases here when changing rehash_table.
bool StringTable::needs_safepoint_rehash_table() {
  // Grow instead of rehash
  if (should_grow()) return false;

  // Already rehashed
  if (_rehashed) return false;

  // Should and can do alt-rehashing
  if (!_needs_rehashing) return false;
  if (!_local_table->is_safepoint_safe()) return false;

  return true;
}
```

Reply:
I prefer using a boolean return value as well. I just had trouble with coming up with a non confusing name for the method. Maybe `needs_safepoint_rehash_table` is good enough.

I'll rewrite the condition with multiple ifs. 

My knowledge of the concurrent hash table is rather limited, but what does the `Should and can do alt-rehashing` comment mean? AFAIK the `_local_table->is_safepoint_safe()` check is just to see if someone is currently resizing the table when we entered the safepoint.

Reply:
I see now, you are describing the negation. `!(!_needs_rehashing || !_local_table->is_safepoint_safe())`
Confused me as the other comments described the condition.
I think I'll rewrite the comment to be more in the same style as the once above.

Reply:
Yeah, feel free. I think the readability improvement is not in comments themselves, but in repeating the basic structure of `rehash_table()` :)